### ISSUE TITLE
L044: CTE in CTE doesn't pass linter 

### ISSUE THREAD

**Issue by pif** [sqlfluff/sqlfluff#1984]
## Observed Behaviour
Rule `L044` fails when there's a CTE inside another CTE.

## Steps To Reproduce

Example SQL:
```sql
with a as (
    with b as (
        select 1
    )

    select * from b
)

select * from a
```

Linter output:
```
== [example/test.sql] FAIL
L:   9 | P:   1 | L044 | Query produces an unknown number of result columns.
```

[Link](https://online.sqlfluff.com/fluffed?sql=H4sIACUNoWEC_yvPLMlQSFRILFbQ4OVSAIJykEASkgAIFKfmpCaXKBhCRDR5uSAMqLCWQlpRfq5CEi8XWAZVNJGXCwBVidBWZAAAAA%3D%3D&dialect=snowflake) to SQLFluff Online output.

## Expected Behaviour
Linter rule `L044` should `PASS` on this, as there's just 1 column expected in the output.

## Debug Output
When ran with `-vvvvv`, here's the output:

```
DEBUG      [L044] Analyzing query: select * from a
DEBUG      [L044] Wildcard: * has target a
DEBUG      [L044] Analyzing query: select 1
DEBUG      [L044] Analyzing query: select * from b
DEBUG      [L044] Wildcard: * has target b
DEBUG      [L044] Query target b is external. Generating warning.
DEBUG      [L044] !! Violation Found: 'Query produces an unknown number of result columns.'
```

**Comment by jpy-git** [comment]
@pif Is this valid SQL?

If it is then nested WITH statements would be a definite antipattern.
```
with a as (
    with b as (
        select 1
    )

    select * from b
)

select * from a
```
should be written
```
with b as (
    select 1
), a as (
    select * from b
)

select * from a
```

**Comment by tunetheweb** [comment]
I'd also say it's an anti-pattern to use the `select *` within both the real statement and CTE a — even if it does ultimately result in a known number of columns at the end. This is a waste of processing effort when potentially an index could be used (though possible some SQL engines will be clever enough to realise this and optimise the query before execution).

Also, consider the following statement:

```sql
with a as (
    select * from b
    UNION ALL
    select * from c
)

select col1 from a
```

There is no guarantee that the `UNION ALL` will continue to work if one of tables b or c change the the number of columns in their table, but the other does not.

So, in my opinion CTEs **should** be checked for rule L044.

To expect L044 to disentangle the CTE and figure out that the original use case does not produce an unknown number of result columns is unrealistic IMHO and, as I say, for index reasons this is an inefficient anti-pattern anyway.

So my vote is to close this issue.

**Comment by jpy-git** [comment]
@tunetheweb I think I agree with you on L044 trying to do too much at the moment, probably better it just focus on flagging wildcards.

for example why introduce all the extra logic to allow:
```sql
with a as (
    select * from b
) as bar

select foo from bar
```
(currently L044 tries to allow for the above)
If it would just be clearer to write
```sql
with a as (
    select foo from b
) as bar

select foo from bar
```

Trying to account for the one simple exception case above opens us up to having to implement the more complicated cases like this issue and the union case you mentioned.

Maybe the best approach is L044 flags any wildcard (regardless of location) and then we could have a config option to `ignore_wildcards` with options `["none", "cte", "subquery", "both"]`. (And if there's one really niche case someone wants they can just use inline ignore, at least we have flagged that they should consider the use of *)

**Comment by jpy-git** [comment]
Also a rule to say "don't use nested WITH statements"

**Comment by barrywhart** [comment]
FWIW, I think the extra complexity is worthwhile, as it gives SQL authors the flexibility to write their code however they want, as long as they don't ultimately have an output table with an unknown number of output columns. This issue sounds like a bug, and I'd be happy to look at it -- I wrote the original L044, partly with the deliberate intent to start adding more sophisticated, complex code analysis to SQLFluff (from a user perspective, this is a good thing!).

**Comment by barrywhart** [comment]
IIRC, this issue arose partly from the fact that SQLFluff probably didn't support nested CTEs at the time L044 was added. (Also, do all dialects support nested CTEs?)

Anyway, I'm 100% in favor of fixing this, very much against reducing functionality of L044 and making it (IMHO) too simplistic and restrictive. I also think this will be a pretty quick fix -- note that L044 already handles recursion (i.e. following a chain of dependent queries); here, we're just adding support for a different kind of recursion (nesting, not just a chain of dependencies in a flat structure).

**Comment by jpy-git** [comment]
Sounds good to me, we can include this and the UNION case in the list of extra cases in #810 to add to L044 then 👍 

I'm going to make a rule regarding the use of nested WITH statements, but would still be good for L044 to handle this case if possible 😄

**Comment by tunetheweb** [comment]
Well won't block it if you want to, but personally I disagree. If we're going to have a rule saying best practice is not to output an unknown number of columns then don't see why best practice would allow them within a statement either 🤷‍♂️

Would be interested to see a real query to see where this would make sense in case I'm missing something.

**Comment by jpy-git** [comment]
@barrywhart could you possibly add a config option `ignore_wildcards` with options like `["none", "cte", "subquery", "both"]` to L044? It can default to the existing behaviour, but then allows for the best of both worlds for those that don't want any wildcards 😄

**Comment by barrywhart** [comment]
Sounds good!  I assigned this one to myself. I've been busy with some personal stuff (plus the Thanksgiving holiday), but hoping to resume work on SQLFluff in the next week or two. 💪🏽

@tunetheweb: It's not absolutely useful/necessary to support this, but remember that data scientists and analysts are not engineers. They will accept a certain set of restrictions, but if something is too restrictive, they'll often quietly rebel by trying to dodge whatever policies you put in place. A good analogy might be password requirements. Many companies are starting to learn that if you make people change passwords too often or make the complexity requirements too high, it'll backfire -- people will start writing their passwords on a sticky note, which is _less_ secure.

I can definitely see refining this and other rules over time, but I want to be cautious about letting our own experience and opinions drive things too much. I don't have numbers, but I would guess that over time, the vast majority of our users will be analysts or data scientists, not engineers. (The usage numbers for Python itself tend to show this.)

**Comment by barrywhart** [comment]
>add a config option ignore_wildcards with options like `["none", "cte", "subquery", "both"]` to L044

@jpy-git: Sure, I'm open to that. As I noted above, the opinion I'm expressing is less "my own" than what I think analysts and data scientists would want. It's a tricky thing, trying to think like someone in another role, but also something I was doing in my day job for the past several years -- I led an ML Platform team, and we didn't get a lot of formal feedback, it was more like being a missionary or anthropologist. 😛

**Comment by tunetheweb** [comment]
I get that, but we offer the option to disable rules so people can configure to their desire if the rules aren't ones they will enforce. Just concerned this will end up with a very complex rule. But maybe I'm overly worrying about a non-issue - let's see what you come up with and then then discuss further then.

**Comment by jpy-git** [comment]
@barrywhart definitely appreciate that, I agree the main demographic of SQL users leans towards the Data Science/Analyst side and can see them wanting this feature, so your points are sensible. I think engineers/developers are often going to be using ORMs rather than raw SQL a lot of the time so def worth not overfitting to us 😄 

My main concern was one of complexity, but I think we're not yet at the level where we need to be too concerned and it sounds like you already have a good plan to implement it so I say go for it (plus a config option to make everyone happy)!

Also happy thanksgiving! 🦃

**Comment by tunetheweb** [comment]
> I think engineers/developers are often going to be using ORMs rather than raw SQL

Don't even start me on those... 🤣

**Comment by barrywhart** [comment]
This fix will add very little complexity.

@NiallRees, can you provide some context here? IIRC, when we were initially adding L044, you noted that dbt has some "best practices" that explicitly **suggest** using `SELECT *` in each model file, I think? Sounds like it'd be worthwhile to find that documented somewhere so we can include a link in the rule docs.

**Comment by pif** [comment]
Hi all, it's great to see this level of engagement! Happy to provide more context.
First of all, the example in the issue was an attempt to create the minimal, reproducible example. 

> Would be interested to see a real query to see where this would make sense in case I'm missing something.

I stumbled upon that when optimising a complex long SQL file, that has multiple steps of preparing data. The 1st level CTEs are the main building blocks of the logic. Some of these 1st level CTEs have smaller multi-step prep that is too small to move out into separate dbt models, but a bit more complex than a a couple joins all done in one query. That was probably the reason for `with` inside the 1st level CTEs.

**Comment by tunetheweb** [comment]
Thanks for the additional context @pif — and most of all, for taking the positives out of above discussion as all too easy to take offence at discussions like this as a personal attack on your issue/code when that is definitely not the intention!

Even if you want rules for the majority of your code base, some of them just won't be enforcable for some queries. Hence why we provide [inline options to disable for those cases](https://docs.sqlfluff.com/en/stable/rules.html#inline-ignoring-errors), and also a [.sqlfluffignore option](https://docs.sqlfluff.com/en/stable/configuration.html#installation-configuration) to exclude a whole directories or individual file without adding messy inline commands.

But let's see what @barrywhart can come up with there. Sounds like he's a plan and he knows SQLFluff better than me so interested to see the results.

**Comment by pif** [comment]
@tunetheweb yeap, eventually excluded that rule from the file, once i verified manually that the code generates a finite # of columns.

Just thought I'd share it here, as the rule's source code does a good job recursing through the statements, just didn't handle that case.

**Comment by barrywhart** [comment]
I have a PR that fixes the issue: #1991. Overall, it was pretty straightforward -- just adding another case to check for and handle nested CTEs.
